Context Improvements

yogeshjain999 opened the pull request:
- Aliasing with indifferent access
- `to_hash` speed improvement - Same-ish as Hash#to_hash
- Support all other `Hash` features on `ctx` object.
- Benchmark checks for performance of indifferent and alias accesses
- Maintain replica for faster access and copy actions

yogeshjain999 committed:
Context Improvements

- Aliasing with indifferent access
- `to_hash` speed improvement - Same-ish as Hash#to_hash
- Support all other `Hash` features (find, dig, collect etc) on `ctx` object.
- Benchmark checks for performance of indifferent and alias accesses
- Maintain replica for faster access and copy actions

yogeshjain999 committed:
Remove ivar of flow_options :tada:

yogeshjain999 commented:
Some `rake benchmark` comparisons

```bash
unknown_read_from_default_hash: 10203086.5 i/s
unknown_read_from_indifferent_hash:  3419061.3 i/s - 2.98x  slower

merge_from_default_hash:   923465.7 i/s
merge_from_indifferent_hash:   176545.1 i/s - 5.23x  slower

read_from_default_hash: 11466571.8 i/s
read_from_indifferent_hash:  8626938.5 i/s - 1.33x  slower

dup_default_hash: 11509089.1 i/s
decompose:  8124801.4 i/s - 1.42x  slower

write_to_default_hash:  8099822.7 i/s
write_to_indifferent_hash:  3712986.5 i/s - 2.18x  slower

delete_from_default_hash:  9736246.3 i/s
delete_from_indifferent_hash:  4511024.0 i/s - 2.16x  slower

dup_default_hash: 12025364.3 i/s
dup_indifferent_hash: 10223733.7 i/s - 1.18x  slower

initialize_default_hash:   610436.1 i/s
initialize_indifferent_hash:   307909.3 i/s - 1.98x  slower

initialize_default_hash:   598107.5 i/s
initialize_aliased_hash:    81555.6 i/s - 7.33x  slower

write_to_default_hash:  8203760.3 i/s
write_to_aliased_hash:  2367645.9 i/s - 3.46x  slower

read_from_default_hash: 11779361.8 i/s
read_from_aliased_hash:  8474189.7 i/s - 1.39x  slower

delete_from_default_hash: 10041370.9 i/s
delete_from_aliased_hash:  2727814.7 i/s - 3.68x  slower

merge_from_default_hash:  1044885.3 i/s
merge_from_aliased_hash:   129195.8 i/s - 8.09x  slower

dup_default_hash: 12364518.6 i/s
dup_aliased_hash: 10541857.4 i/s - 1.17x  slower
```

yogeshjain999 committed:
Override inspect in aliasing extension

yogeshjain999 committed:
Define Container::WithAliases instead of creating anonymous class

yogeshjain999 committed:
Change `dup` to `to_hash` in benchmark naming, `dup` was bit confusing

apotonick commented on lib/trailblazer/context.rb:
```diff
@@ -12,23 +12,22 @@ module Trailblazer
 
-    def build(wrapped_options, mutable_options, container_class: Context::Container, **flow_options)
-      if flow_options[Context::Extension::Aliasing::CONFIG_KEY]
-        container_class = Class.new(container_class).prepend(Context::Extension::Aliasing)
```
Why don't we inject the container/ctx class via `flow_options`?

yogeshjain999 replied:
Well, not a biggie but few things that I thought about this,

- "container_class" seemed kind of generic to pass in `flow_options`. It would be cool if we could have a structure like `flow_options: { context: { container_class: Context::Container, aliases: { ... } } }`

- Above check to replace "container_class" with `Context::WithAliases` when `context_alias` are present might have led to confusion.

- Which use case there could be to pass containers apart from what we've defined ?

apotonick replied:
It's always better to avoid globals and pass options around programmatically, and since we're already in a functional flow, I think it's less error prone (always imagine you're too test this behavior and have to override a global for a certain time, only. Seen this everywhere in Rails :cry: ).

I like the structure for the options you propose!

yogeshjain999 replied:
Cool bro. If we have that namespace for `context options`, then it's sorted!

yogeshjain999 committed:
Namespace context related options in flow_options

```ruby
{ flow_options: { context_options: { aliases: {}, ** } } }
```

yogeshjain999 committed:
Autoload default IndifferentAccess replica class

- Replica class can be configured to not do indifferent access and
increase little performance
- Add test for passing custom replica class

apotonick commented on lib/trailblazer/context.rb:
```diff
@@ -11,23 +11,17 @@ module Trailblazer
-      return with_aliases if flow_options.key?(:context_alias)
-      return default
+    def build(wrapped_options, mutable_options, context_options: {}, **)
+      klass = context_options[:container_class] || Container
```
I'm curious if we need the defaulting here - I'd love to have defaulting for all constants happen as "highly" as possible so maybe it's something the user has to take care about?

yogeshjain999 replied:
uhmm didn't this exactly. Do you mean user should pass `container_class` always ?

yogeshjain999 committed:
Make no assumptions about container_class & replica_class

Explicitly passing all values make ctx.build API cleaner

yogeshjain999 committed:
No need to pass empty context_options to `Context()`

yogeshjain999 committed:
Add safeguard to check container_class when passing aliases
